### DISCUSSION TITLE
Compose data in vue by using `useQuery` data in a Pinia store?

### DISCUSSION THREAD

**Discussion by d1gitalrain** [TanStack/query#6798] [Q&A] [answered]
Hey all!

I'm trying to add a layer (in the form of a Pinia store) between the `useQuery` for an object and the handful of components that use it. The reason for this Pinia store is that the page is quite complicated and has a bunch of components that interact with each other, so there is an overall "page" store that handles interfacing between all the components. I'd love to integrate vue-query into this so I can make use of caching and some of the other features, but I'm struggling to get it to work.

I understand using `useQuery` outside of setup or injectable context might break some reactivity, but I'm looking for some help/ideas towards what I'm trying to achieve. Here's what I've tried so far:

```ts
// api/todos.ts

// Wrapping query in a composable...
// Maybe I could put some computed's in here, but I'd like to keep this data layer separate from page logic
export function useTodos() {
    const { data: todos } = useQuery({ queryKey: ["todos"], queryFn: fetchTodos });

    return { todos };
}
```

And the page store:

```ts
// stores/page.ts

export const usePageStore = defineStore("page", () => {
    const { todos } = useTodos();

    const todosByUserId = computed(() => {
        return _.groupBy(todos.value, "userId");
    });

    // lots of other getters...
    // lots of other actions...
    
    return {
        todosByUserId,

        // ...
    }
});

```

Then, my page is a composite of a bunch of components that each use different parts (some overlap) of the store:

```vue
<!-- Page.vue -->

<template>
    <TodoMenuBar />    <!-- Uses entirety of the "todos" data to show options/filters -->
    <TodoList />       <!-- Uses entirety of the "todos" data to show a list -->
    <TodoDetailsSideBar />    <!-- Can show a single todo item -->
    <!-- More components... -->
</template>
```

The problem is, the `todosByUserId` computed does not reactively update to when the `todos` query finishes, it just remains empty. However, if I move that computed value into one of the components themselves, it works. Is `useQuery` not meant to be used like this, or am I missing something reactivity-wise? Composing complex data into shared stores seems natural in my head, but let me know if I'm missing something here. Thanks!

**Comment by Mini-ghost** [comment]
Hi! Can you provide a minimal reproduction for this problem?

**Comment by Filip1139** [comment]
Do You find solution? I have same problem

**Comment by ikliko** [comment]
I think your code is related to reactivity and lifecycle hooks in Vue and Pinia. Also `vue-query` have specifications. I will give you some code examples below.

api/todos.ts
```
import { useQuery } from 'vue-query';

export function useTodos() {
    const { data: todos, isLoading, isError } = useQuery({
        queryKey: ["todos"],
        queryFn: fetchTodos
    });

    return { todos, isLoading, isError };
}
```

stores/page.ts
```
import { defineStore } from 'pinia';
import { ref, computed, watchEffect } from 'vue';
import { useTodos } from '../api/todos';

export const usePageStore = defineStore('page', () => {
    const { todos, isLoading, isError } = useTodos();

    const todosRef = ref([]);
    const isLoadingRef = ref(true);
    const isErrorRef = ref(false);

    watchEffect(() => {
        if (todos.value) {
            todosRef.value = todos.value;
        }
        isLoadingRef.value = isLoading.value;
        isErrorRef.value = isError.value;
    });

    const todosByUserId = computed(() => {
        return _.groupBy(todosRef.value, 'userId');
    });

    return {
        todosRef,
        isLoadingRef,
        isErrorRef,
        todosByUserId
    };
});
```

Page.vue
```
<template>
    <TodoMenuBar />
    <TodoList />
    <TodoDetailsSideBar />
    <!-- More components... -->
</template>

<script setup>
import { usePageStore } from '../stores/page';

const pageStore = usePageStore();
</script>
```


TodoMenuBar.vue
```
<template>
    <div v-if="isLoading">Loading...</div>
    <div v-else>
        <!-- more code  -->
    </div>
</template>

<script setup>
import { computed } from 'vue';
import { usePageStore } from '../stores/page';

const pageStore = usePageStore();
const todos = computed(() => pageStore.todosRef);
const isLoading = computed(() => pageStore.isLoadingRef);
const todosByUserId = computed(() => pageStore.todosByUserId);
</script>
```

**Comment by d1gitalrain** [comment, marked as answer]
Sorry, forgot about this thread. I did end up figuring out the issue. [Here's](https://codesandbox.io/p/sandbox/pinia-vue-3-demo-forked-s6693v?file=%2Fsrc%2Fquery%2Ftodos.ts%3A15%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxl2pool00062a69ypvpoqpq%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxl2pool00022a693vbxsa90%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxl2pool00032a69gis509co%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxl2pool00052a69kevg39ot%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxl2pool00022a693vbxsa90%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxl2pool00012a695n3iqeuh%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxl2rj0200302a691tnwh585%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A47%252C%2522startColumn%2522%253A9%252C%2522endLineNumber%2522%253A47%252C%2522endColumn%2522%253A9%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.vue%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxl3jdsx00022a69bdwnb3kj%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A15%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A15%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fquery%252Ftodos.ts%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clxl2pool00022a693vbxsa90%2522%252C%2522activeTabId%2522%253A%2522clxl3jdsx00022a69bdwnb3kj%2522%257D%252C%2522clxl2pool00052a69kevg39ot%2522%253A%257B%2522id%2522%253A%2522clxl2pool00052a69kevg39ot%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxl2pool00042a69mnj9gtcf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clxl2pool00042a69mnj9gtcf%2522%257D%252C%2522clxl2pool00032a69gis509co%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clxl2pool00032a69gis509co%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) a min example if anyone is stuck on this.

(Not sure what the comment above me is trying to say. I wouldn't recommend using `watchEffect` to manage values like they are doing. Passing around refs is just fine.)
